revocation/ocsp/errors.go

```diff
@@ -0,0 +1,63 @@
+// Package OCSP provides methods for checking the OCSP revocation status of a
+// certificate chain, as well as errors related to these checks
+package ocsp
+
+import (
+	"fmt"
+	"time"
+)
+
+// RevokedError is returned when the certificate's status for OCSP is
+// ocsp.Revoked
+type RevokedError struct{}
+
+func (e RevokedError) Error() string {
+	return "certificate is revoked via OCSP"
+}
+
+// UnknownStatusError is returned when the certificate's status for OCSP is
+// ocsp.Unknown
+type UnknownStatusError struct{}
+
+func (e UnknownStatusError) Error() string {
+	return "certificate has unknown status via OCSP"
+}
+
+// OCSPCheckError is returned when there is an error during the OCSP revocation
+// check, not necessarily a revocation
+type OCSPCheckError struct {
+	Err error
+}
+
+func (e OCSPCheckError) Error() string {
+	msg := "error checking revocation status via OCSP"
+	if e.Err != nil {
+		return fmt.Sprintf("%s: %s", msg, e.Err.Error())
+	}
+	return msg
+}
+
+// NoOCSPServerError is returned when the OCSPServer is not specified.
+type NoOCSPServerError struct{}
+
+func (e NoOCSPServerError) Error() string {
+	return "no valid OCSPServer found"
+}
+
+// PKIXNoCheckError is returned when the OCSP signing cert is missing the
+// id-pkix-ocsp-nocheck extension.
+type PKIXNoCheckError struct{}
+
+func (e PKIXNoCheckError) Error() string {
+	return fmt.Sprintf("an OCSP signing cert is missing the id-pkix-ocsp-nocheck extension (%s)", pkixNoCheckOID)
+}
```

Comment:
> when id-pkix-ocsp-nocheck is not found that it uses a ServerResult with an error of nil and a result of NonRevocable

This is not correct because technically the signing certificate is revocable using ocsp but the certificate which signed ocsp is non-revocable.

>  I think we should consider changing the name of NonRevocable to CRLRequired

Why should we do that? IMO if we are not showing `CRLRequired` when ocsp is not configured on CRL then we shouldn't do it for this usecase as both are analogous.

> Additionally, since this PR does not incorporate CRL checking, should we be differentiating between the scenarios you provided?
For current implementation we shouldn't as we don't support CRL.

Reply:
The above makes sense. Based on the spec, my understanding is if the extension is not present that we will only be conducting CRL checks to verify OCSP responders. As a result, we need to determine how to handle these situations where the extension is not present (while we are not supporting CRL). I see three potential options:
 1. If the extension is missing, use an error of PKIXNoCheckError and a Result of Unknown. This would essentially say that this response is unusable and there is an error. I think this best aligns with the situation, since we currently don't support checking responders that don't have this result. We can later update this ServerResult once we add CRL to use the error/result of the CRL check for the responder cert.
 2. If the extension is missing, use an error of PKIXNoCheckError and create a new result only for these errors. This would not necessarily make this an error, only an indication that it needs to be checked further. We could handle this result differently in notation-go if we wanted to support responses that don't have the extension. However, I think supporting such extensions may be a potential risk.
 3. Completely ignore this extension and treat the response as valid. This would mean that we aren't validating responders, which I think is a potential risk, which I believe makes this a nonstarter.

I would suggest option 1 (I've ordered them based on my preferences), do you agree with this? If not, what would you suggest is the best behavior here?

Reply:
Consensus is to go with option 3, as it best matches other cases where CRL check should be done but is missing